main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala

```diff
@@ -0,0 +1,454 @@
+    }
+
+    ("status" -> "success") ~ ("result" -> CompletionList(isIncomplete = true, completions).toJSON)
+  }
+
+  private def getCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): List[CompletionItem] = {
+    getKeywordCompletions() ++
+      getSnippetCompletions() ++
+      getVarCompletions() ++
+      getDefAndSigCompletions() ++
+      getWithCompletions() ++
+      getInstanceCompletions()
+  }
+
+  private def keywordCompletion(name: String)(implicit context: Context, index: Index, root: TypedAst.Root): CompletionItem = {
+    CompletionItem(label = name,
+      sortText = "9" + name,
+      textEdit = TextEdit(context.range, name),
+      kind = CompletionItemKind.Keyword)
+  }
+
+  private def getKeywordCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): List[CompletionItem] = {
+    // TODO: keyword-specific help text?
+    // NB: Please keep the list alphabetically sorted.
+    List(
+      "@benchmark", 
+      "@Deprecated", 
+      "@Experimental", 
+      "@Parallel", 
+      "@ParallelWhenPure", 
+      "@Lazy", 
+      "@LazyWhenPure", 
+      "@Space", 
+      "@test", 
+      "@Time",
+      "and", 
+      "as", 
+      "case", 
+      "chan", 
+      "choose", 
+      "class", 
+      "def", 
+      "deref", 
+      "discard", 
+      "do", 
+      "eff", 
+      "else", 
+      "enum", 
+      "false", 
+      "fix", 
+      "forall",
+      "force", 
+      "from", 
+      "get", 
+      "if", 
+      "import", 
+      "Impure", 
+      "instance", 
+      "into", 
+      "lat", 
+      "law", 
+      "lazy", 
+      "let", 
+      "match", 
+      "namespace", 
+      "new",
+      "not", 
+      "null", 
+      "opaque", 
+      "or", 
+      "override", 
+      "project", 
+      "pub", 
+      "Pure", 
+      "query", 
+      "Record", 
+      "ref", 
+      "region", 
+      "rel", 
+      "Schema", 
+      "sealed",
+      "select", 
+      "set", 
+      "solve", 
+      "spawn", 
+      "true", 
+      "try", 
+      "type", 
+      "use", 
+      "where", 
+      "with"
+    ) map keywordCompletion
+  }
+
+  private def snippetCompletion(name: String, snippet: String, documentation: String)(implicit context: Context, index: Index, root: TypedAst.Root): CompletionItem = {
+    CompletionItem(label = name,
+      sortText = "8" + name,
+      textEdit = TextEdit(context.range, snippet),
+      documentation = Some(documentation),
+      insertTextFormat = InsertTextFormat.Snippet,
+      kind = CompletionItemKind.Snippet)
+  }
+
+  private def getSnippetCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): List[CompletionItem] = {
+    List(
+      // NB: Please keep the list alphabetically sorted.
+      snippetCompletion("main",
+        "def main(): Unit & Impure = \n    println(\"Hello World!\")",
+        "snippet for Hello World Program"),
+      snippetCompletion("query",
```

Comment:
In the future, I guess we should decide if snippets even make sense!

Reply:
Agreed. Ill leave it like this for now though?